### PR TITLE
Fix phone unique constraints

### LINKED ISSUE #20195: Unique constraint on phone type field falsely throws an error
**Scenario:**
1. Log in
2. Go to Settings > Data model > People > Phones field
3. Toggle Unique on and save
4. Go to People view
5. Insert the same number in 2 records with different landline code (e.g. +1 123456789 and +32 123456789)

**Actual:** Error is thrown as phone number is the same despite different landline code

**Expected:** No error is thrown as validation checks both landline code as well as phone number itself

### PR DESCRIPTION
**Pull request by MkDev11**
## Summary

Closes #20195

Fix phone field unique constraints so phone numbers are considered unique by both `primaryPhoneNumber` and `primaryPhoneCallingCode`.

- Include `primaryPhoneCallingCode` in the shared phone composite unique constraint metadata
- Align the frontend settings composite field config with the backend metadata
- Return all included unique composite subfields when building create-many conflict fields
- Match composite unique conflict fields as a group during create-many upserts

## Root Cause

Phone composite metadata only marked `primaryPhoneNumber` as part of the unique constraint. That made different international phone numbers with the same national number conflict, for example `+1 123456789` and `+32 123456789`.

## Test Plan

- `yarn workspace twenty-shared build`
- `jest --runTestsByPath <index action handler and create-many utility specs>`
- `prettier --check <touched files>`
- `oxlint --type-aware <touched files>`
- `nx run twenty-shared:typecheck`
- `nx run twenty-server:typecheck`
- `nx run twenty-front:typecheck`

### CODE REVIEW

**Comment by MkDev11** [comment]
@charlesBochet can you please review the pr?

**Comment by MkDev11** [comment]
the API changes report appears to be a tooling/baseline introspection parse issue
the `api-breaking-changes` check passed, and this PR does not change GraphQL/OpenAPI schema definitions

**Review by etiennejouan — COMMENTED** [review]
LGTM - Thanks for your contribution !
Could you, please, add integration tests here : packages/twenty-server/test/integration/graphql/suites/unique-field/unique-phones-field-null-equivalence.integration-spec.ts :
-> 'should allow creating when same phoneNumber but different phoneCallingCode'

**Comment by MkDev11** [comment]
> LGTM - Thanks for your contribution ! Could you, please, add integration tests here : packages/twenty-server/test/integration/graphql/suites/unique-field/unique-phones-field-null-equivalence.integration-spec.ts : -> 'should allow creating when same phoneNumber but different phoneCallingCode'

@etiennejouan please check the update

**Comment by etiennejouan** [comment]
Have you test the PR locally ? Via UI, I can't create a unique phone field.
I have issue with defaultValue. And we're not totally sure of the implementation. Currently we don't accept defaultValue for unique field, metadata migration engine may need to be updated, not trivial.
This phone composite isIncludedInUniqueConstraint update would also need a migration command for all existing unique phone field indexes.

Closing this PR. Feel free to reach me on discord to discuss implem. Or re-open directly with update

**Comment by mkdev5** [comment]
@etiennejouan can you reopen the PR

**Comment by mkdev5** [comment]
@etiennejouan can you review the update?

**Comment by mkdev5** [comment]
I tested the fix locally and it does work correctly

**Comment by Ella-horizon** [comment]
@etiennejouan could you give me your feedback, please?

**Comment by ClaudeCenter** [comment]
Hi @etiennejouan how are you?
I am going to contribute to `twenty` project. Now this PR was re-opened twice and drafted. This means the issue is still vailid.
I think I can complete this PR correctly. Can I pick this up?

**Comment by etiennejouan** [comment]
> Hi @etiennejouan how are you? I am going to contribute to `twenty` project. Now this PR was re-opened twice and drafted. This means the issue is still vailid. I think I can complete this PR correctly. Can I pick this up?

Hi @ClaudeCenter, thanks for your motivation. This one is too complex for a first issue and fixing is in progress. Please, feel free to take an other issue

File: packages/twenty-front/index.html [hunk 1]
```diff
@@ -41,7 +41,8 @@
     <!-- BEGIN: Twenty Config -->
     <script id="twenty-env-config">
       window._env_ = {
-        // This will be overwritten
+        // This will be overwritten in production. In dev, Vite substitutes %VAR% tokens from .env.
+        REACT_APP_SERVER_BASE_URL: '%REACT_APP_SERVER_BASE_URL%',
```
Comment (prastoin) [inline packages/twenty-front/index.html]: I don't understand the motivation here

File: packages/twenty-server/src/database/commands/upgrade-version-command/2-4/2-4-workspace-command-1778000000000-rebuild-unique-phone-indexes.command.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+import { type FlatIndexMetadata } from 'src/engine/metadata-modules/flat-index-metadata/types/flat-index-metadata.type';
+import { WorkspaceSchemaManagerService } from 'src/engine/twenty-orm/workspace-schema-manager/workspace-schema-manager.service';
+import { WorkspaceCacheService } from 'src/engine/workspace-cache/services/workspace-cache.service';
+import {
+  createIndexInWorkspaceSchema,
+  dropIndexFromWorkspaceSchema,
+} from 'src/engine/workspace-manager/workspace-migration/workspace-migration-runner/action-handlers/index/utils/index-action-handler.utils';
+import { getWorkspaceSchemaName } from 'src/engine/workspace-datasource/utils/get-workspace-schema-name.util';
+
+@RegisteredWorkspaceCommand('2.4.0', 1778000000000)
```
Comment (prastoin) [inline packages/twenty-server/src/database/commands/upgrade-version-command/2-4/2-4-workspace-command-1778000000000-rebuild-unique-phone-indexes.command.ts]: we should move all commands to 2.5.0

File: packages/twenty-server/src/engine/metadata-modules/flat-field-metadata/utils/normalize-composite-default-value.util.ts [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+      if (blocknote === null && markdown === null) {
+        return null;
+      }
+
+      return { blocknote, markdown };
+    }
+
+    default:
+      assertUnreachable(type);
+  }
```
Comment (prastoin) [inline packages/twenty-server/src/engine/metadata-modules/flat-field-metadata/utils/normalize-composite-default-value.util.ts]: ***Remark:***
Feel like we could also split in several dedicated utils in order to downscope
Naming isn't great too

File: packages/twenty-server/src/engine/metadata-modules/flat-field-metadata/utils/normalize-composite-default-value.util.ts [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+};
```
Comment (prastoin) [inline packages/twenty-server/src/engine/metadata-modules/flat-field-metadata/utils/normalize-composite-default-value.util.ts]: ***Remark:*** We're introducing a new normalization that should also be applied to existing records for all related metadata type ?

File: packages/twenty-server/src/engine/metadata-modules/flat-field-metadata/utils/get-default-flat-field-metadata-from-create-field-input.util.ts [hunk 5]
```diff
@@ -42,7 +44,12 @@ export const getDefaultFlatFieldMetadata = ({
     type: createFieldInput.type,
     universalIdentifier: createFieldInput.universalIdentifier ?? v4(),
     options: createFieldInput.options ?? null,
-    defaultValue: defaultValue ?? generateDefaultValue(createFieldInput.type),
+    defaultValue: isCompositeFieldMetadataType(createFieldInput.type)
+      ? normalizeCompositeDefaultValue(
+          defaultValue ?? generateDefaultValue(createFieldInput.type),
+          createFieldInput.type,
+        )
+      : (defaultValue ?? generateDefaultValue(createFieldInput.type)),
```
Comment (prastoin) [inline packages/twenty-server/src/engine/metadata-modules/flat-field-metadata/utils/get-default-flat-field-metadata-from-create-field-input.util.ts]: Looks redundant could have a better abstraction that handles this gym

File: packages/twenty-server/src/engine/workspace-manager/workspace-migration/workspace-migration-builder/validators/services/flat-index-metadata-validator.service.ts [hunk 6]
```diff
@@ -135,6 +137,12 @@ export class FlatIndexValidatorService {
               userFriendlyMessage: msg`Field referenced in index does not exist`,
             });
           } else {
+            const compositeType = isCompositeUniversalFlatFieldMetadata(
+              relatedFlatField,
+            )
+              ? compositeTypeDefinitions.get(relatedFlatField.type)
+              : undefined;
+
```
Comment (prastoin) [inline packages/twenty-server/src/engine/workspace-manager/workspace-migration/workspace-migration-builder/validators/services/flat-index-metadata-validator.service.ts]: ***Polish:*** Could be retrieved only when isUnique true

File: packages/twenty-server/src/engine/workspace-manager/workspace-migration/workspace-migration-runner/action-handlers/index/utils/__tests__/compute-flat-index-field-column-names.util.spec.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import { FieldMetadataType } from 'twenty-shared/types';
+
+import { type MetadataFlatEntityMaps } from 'src/engine/metadata-modules/flat-entity/types/metadata-flat-entity-maps.type';
+import { type FlatFieldMetadata } from 'src/engine/metadata-modules/flat-field-metadata/types/flat-field-metadata.type';
+import { type FlatIndexFieldMetadata } from 'src/engine/metadata-modules/flat-index-metadata/types/flat-index-metadata.type';
+import { computeFlatIndexFieldColumnNames } from 'src/engine/workspace-manager/workspace-migration/workspace-migration-runner/action-handlers/index/utils/index-action-handler.utils';
+
+describe('computeFlatIndexFieldColumnNames', () => {
```
Comment (prastoin) [inline packages/twenty-server/src/engine/workspace-manager/workspace-migration/workspace-migration-runner/action-handlers/index/utils/__tests__/compute-flat-index-field-column-names.util.spec.ts:8]: ***Remark:*** unit test file is too global whereas it's only coverage specific call to `computeFlatIndexFieldColumnNames`
Should be a dedicated unit test suite of this util instead
```
      computeFlatIndexFieldColumnNames({
        flatIndexFieldMetadatas,
        flatFieldMetadataMaps,
      }),
```

File: packages/twenty-server/src/engine/workspace-manager/workspace-migration/workspace-migration-runner/utils/__tests__/generate-column-definitions.util.spec.ts [hunk 8]
```diff
@@ -246,6 +246,94 @@ describe('Generate Column Definitions', () => {
...
+          expect.objectContaining({
+            name: 'phonePrimaryPhoneCallingCode',
+            default: 'NULL',
+          }),
+          expect.objectContaining({
+            name: 'phoneAdditionalPhones',
+            default: 'NULL',
+          }),
+        ]),
+      );
```
Comment (prastoin) [inline packages/twenty-server/src/engine/workspace-manager/workspace-migration/workspace-migration-runner/utils/__tests__/generate-column-definitions.util.spec.ts]: ***remark:*** Unless I'm mistaken could be replaced by a strict equal instead

File: packages/twenty-server/src/engine/metadata-modules/flat-field-metadata/utils/normalize-composite-default-value.util.ts [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+
+import { isNullEquivalentArrayFieldValue } from 'src/engine/api/common/common-args-processors/data-arg-processor/utils/is-null-equivalent-array-field-value.util';
+import { CompositeFieldMetadataType } from 'src/engine/metadata-modules/field-metadata/types/composite-field-metadata-type.type';
+
+const isNullEquivalentTextDefaultValue = (value: unknown): boolean =>
+  value === "''" || value === '';
+
+export const normalizeCompositeDefaultValue = (
+  defaultValue: FieldMetadataDefaultValueForAnyType,
+  type: CompositeFieldMetadataType,
```
Comment (prastoin) [inline packages/twenty-server/src/engine/metadata-modules/flat-field-metadata/utils/normalize-composite-default-value.util.ts]: record's arg

File: packages/twenty-server/test/integration/graphql/suites/unique-field/unique-phones-field-null-equivalence.integration-spec.ts [hunk 10]
```diff
@@ -9,6 +9,9 @@ import { FieldMetadataType } from 'twenty-shared/types';
 
 const OBJECT_SINGULAR = 'uniquePhonesTestObject';
 const OBJECT_PLURAL = 'uniquePhonesTestObjects';
+// Custom object metadata maps to a physical table prefixed with `_`
+// (see computeTableName in src/engine/utils/compute-table-name.util.ts).
```
Comment (prastoin) [inline packages/twenty-server/test/integration/graphql/suites/unique-field/unique-phones-field-null-equivalence.integration-spec.ts]: ***Polish:*** depends if it's system or not
please remove comment

File: packages/twenty-server/test/integration/graphql/suites/unique-field/unique-phones-field-null-equivalence.integration-spec.ts [hunk 11]
```diff
@@ -104,23 +113,86 @@ describe('unique PHONES field with empty values', () => {
...
+    const secondResponse = await createOneOperation({
+      objectMetadataSingularName: OBJECT_SINGULAR,
+      input: {
+        id: secondId,
+        [FIELD_NAME]: {
+          primaryPhoneNumber: '4155552671',
+          primaryPhoneCallingCode: '+32',
+          primaryPhoneCountryCode: 'BE',
+        },
+      },
```
Comment (prastoin) [inline packages/twenty-server/test/integration/graphql/suites/unique-field/unique-phones-field-null-equivalence.integration-spec.ts:184]: ***Praise:*** Overall nice additions here

**Review by prastoin — COMMENTED** [review]
Solid work over there !
Pretty complex to be handled
Refactored the format returned by the `getConflictingField` as the `getMatchingRecordId` was handling back and forth redundant data transpilation

We should double check we're not applying new constraint to non backfilled metadata field type

**Comment by mkdev5** [comment]
@prastoin @etiennejouan addressed all comments and pushed the update

File: packages/twenty-server/src/database/commands/upgrade-version-command/2-5/2-5-workspace-command-1778000001000-normalize-composite-field-defaults.command.ts [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+    for (const { tableName, columnName } of backfillTargets) {
+      await dataSource.query(
+        `UPDATE "${schemaName}"."${tableName}"
+         SET "${columnName}" = NULL
+         WHERE "${columnName}"::text IN ('', '""')`,
+        undefined,
+        undefined,
+        { shouldBypassPermissionChecks: true },
+      );
+
```
Comment (prastoin) [inline packages/twenty-server/src/database/commands/upgrade-version-command/2-5/2-5-workspace-command-1778000001000-normalize-composite-field-defaults.command.ts]: I'm quite afraid of this

File: packages/twenty-server/src/engine/metadata-modules/flat-field-metadata/utils/sanitize-raw-update-field-input.ts [hunk 13]
```diff
@@ -45,6 +47,17 @@ export const sanitizeRawUpdateFieldInput = ({
...
 
+  if (
+    updatedEditableFieldProperties.defaultValue !== undefined &&
+    isCompositeFieldMetadataType(existingFlatFieldMetadata.type)
+  ) {
+    updatedEditableFieldProperties.defaultValue =
+      nullifyEmptyCompositeDefaultValue({
+        defaultValue: updatedEditableFieldProperties.defaultValue,
+        fieldType: existingFlatFieldMetadata.type,
+      });
```
Comment (prastoin) [inline packages/twenty-server/src/engine/metadata-modules/flat-field-metadata/utils/sanitize-raw-update-field-input.ts:58]: lets not care here but from my understanding we could create a generic util that set as null any undefined or `''` or `"''"` composite field default value properties instead of having n mappers ?

**Review by prastoin — APPROVED** [review]
Cool !  this will require a strong QA

**Comment by mkdev5** [comment]
@etiennejouan thanks for your help!
